### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Roma-Koval**

### CODE REVIEW

File: app/src/main/java/com/example/giphyservice/data/repository/CustomCallback.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.example.giphyservice.data.repository
+
+import com.example.giphyservice.data.model.Gif
+
+interface CustomCallback {
```
Comment (PavloZoria) [inline app/src/main/java/com/example/giphyservice/data/repository/CustomCallback.kt]: give better naming for that callback.
Improvement: you can make it generic, https://stackoverflow.com/a/15260988/8295472

File: app/src/main/java/com/example/giphyservice/ui/CloseableCoroutineScope.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import java.io.Closeable
+import kotlin.coroutines.CoroutineContext
+
+
+class CloseableCoroutineScope(
```
Comment (PavloZoria) [inline app/src/main/java/com/example/giphyservice/ui/CloseableCoroutineScope.kt]: why do you need that?

File: app/src/main/java/com/example/giphyservice/ui/MainActivity.kt [hunk 3]
```diff
@@ -17,20 +22,44 @@ class MainActivity : AppCompatActivity() {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
 
-        val gifsAdapter = GifsAdapter(context = this, mListener = object : GifsAdapter.OnItemClickListener {
```
Comment (PavloZoria) [inline app/src/main/java/com/example/giphyservice/ui/MainActivity.kt:20]: `private var viewModel: MainViewModel = MainViewModel()` is incorrect. read documentation about viewModel again

File: app/src/main/java/com/example/giphyservice/ui/MainActivity.kt [hunk 4]
```diff
@@ -17,20 +22,44 @@ class MainActivity : AppCompatActivity() {
...
+
         viewModel.objectData.observe(this) { gifs ->
             gifsAdapter.updateGifs(gifs)
+            textView.visibility = View.GONE
+            errorButton.visibility = View.GONE
+        }
+
+        viewModel.error.observe(this) { error ->
+            textView.text = error?.message
+            textView.visibility = View.VISIBLE
```
Comment (PavloZoria) [inline app/src/main/java/com/example/giphyservice/ui/MainActivity.kt]: refactor to `isVisible`

Comment (PavloZoria) [reply app/src/main/java/com/example/giphyservice/ui/MainActivity.kt]: everywhere

File: app/src/main/java/com/example/giphyservice/ui/MainViewModel.kt [hunk 5]
```diff
@@ -4,35 +4,37 @@ import androidx.lifecycle.MutableLiveData
...
+// The ViewModel retrieves data from the Model and exposes it to the View through observable properties or LiveData objects.
 // It also communicates with the Model to update data based on user actions
+//
+// The ViewModel exists from when you first request a ViewModel until the activity is finished and destroyed.
+// CoroutineScope automatically follows the ViewModel's lifecycle.
+// Closeable object in the ViewModel's constructor  automatically closes when the ViewModel instance is cleared.
 
-class MainViewModel : ViewModel() {
+class MainViewModel(
+    coroutineScope: Closeable = CloseableCoroutineScope()
```
Comment (PavloZoria) [inline app/src/main/java/com/example/giphyservice/ui/MainViewModel.kt]: I don't think you need a coroutine scope here. You don't use it

File: app/src/main/java/com/example/giphyservice/ui/list/GifsAdapter.kt [hunk 6]
```diff
@@ -13,7 +13,7 @@ class GifsAdapter(
     var mListener: OnItemClickListener,
     val context: Context,
 ) : RecyclerView.Adapter<GifViewHolder>() {
-    val gifs: MutableList<Gif> = mutableListOf()
+    private val gifs: MutableList<Gif> = mutableListOf()
 
     fun updateGifs(gif: List<Gif>) {
         gifs.clear()
```
Comment (PavloZoria) [inline app/src/main/java/com/example/giphyservice/ui/list/GifsAdapter.kt:17]: about: 
```
GifViewHolder(
            LayoutInflater.from(context).inflate(R.layout.item_layout, parent, false), mListener
        )
```
you can get context from the `parent`

File: app/src/main/res/layout/activity_main.xml [hunk 7]
```diff
@@ -12,4 +13,43 @@
...
+        app:layout_constraintTop_toTopOf="@+id/recyclerView"
+        android:layout_marginHorizontal="24dp"
+        android:textAlignment="center"
+        android:visibility="gone"
+        tools:visibility="visible"
+        tools:text="@string/tools_Text_in_TextView"
+        android:textColor="@color/red"/>
+
+    <Button
+        android:visibility="gone"
```
Comment (PavloZoria) [inline app/src/main/res/layout/activity_main.xml]: fix codestyle(rearrange properties)

File: app/src/main/java/com/example/giphyservice/ui/DetailFragment.kt [hunk 8]
```diff
@@ -4,42 +4,43 @@ import android.os.Bundle
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
+import androidx.core.os.bundleOf
 import androidx.fragment.app.Fragment
 import com.bumptech.glide.Glide
 import com.example.giphyservice.databinding.FragmentDetailBinding
 
+private const val URL_KEY = "url"
```
Comment (PavloZoria) [inline app/src/main/java/com/example/giphyservice/ui/DetailFragment.kt]: move them into the companion object

File: app/src/main/java/com/example/giphyservice/ui/DetailFragment.kt [hunk 9]
```diff
@@ -4,42 +4,43 @@ import android.os.Bundle
...
 
-        val url = arguments?.getString("url")
+        val url = arguments?.getString(URL_KEY)
         Glide.with(this@DetailFragment).load(url).into(binding.fragmentImageView)
 
-/*        setFragmentResultListener("requestKey") { requestKey, bundle ->
-            val url = bundle.getString("url")
-            Glide.with(this@DetailFragment).load(url).into(getBinding.fragmentImageView)
-        }*/
+        /*        setFragmentResultListener("requestKey") { requestKey, bundle ->
```
Comment (PavloZoria) [inline app/src/main/java/com/example/giphyservice/ui/DetailFragment.kt]: clean code

Comment (PavloZoria) [reply app/src/main/java/com/example/giphyservice/ui/DetailFragment.kt]: everywhere

Comment (Roma-Koval) [reply app/src/main/java/com/example/giphyservice/ui/DetailFragment.kt]: done!

**Pull request merged**